README.md
APCS final project


initially this whole game was supposed to be much more like doodlejump where you moved up the game screen and went through infinitely randomly spawning platforms and faced enemies in varying difficulties. however by the time i had a page full in my composition notebook trying to find an optimized algorithm for procedural generation of platform size, platform length, enemies, etc. such that the game was A) always possible to beat and B) fun, i figured we were already kind of late and should start getting stuff on screen. we then kept building on "getting stuff on screen" until we got where we are now, which is  :

------- 
GAME DETAILS
============
youre a colorful but decidedly aloof character in a bright, sunny world. you can only move left and right (your legs only move too, your arms and head motionless as you shuffle along), and flip your gravity (which is much more kickass than jumping). all of a sudden a cluster of angry dark black red clouds, so angry they vigorously shake to and fro in the air out of anger, so angry theyre willing to chase you down (albeit pretty slowly cause theyre so busy being angry) and TAKE YOUR MONEY OUT OF YOUR WALLET! (which you only have $1k of anyways)! Thankfully you have DEADLY FEET which, on contact with an enemy, takes your money AND THEIRS back (you get a lot more back from enemies than they from you, i assume they scare the dollars into reproducing), kills the enemy (but prompts them to send a last dying message to their cluster, which in turn causes an enemy to spawn anew somewhere else). but if you contact them with anything except your deadly feet, they take away your money! and be careful, because the contrast between the ultra-sachharine world and the evil, angsty clouds greatly confuses our protagonist, limiting his lifespan to just 300 seconds! gain as much money from your enemies as possible before your inevitable death! 

-------

we were planning to add a punching feature but there was not enough time to fully integrate that. 
what we were working on is in the "punch repo". we were also planning to :

- make the platform more intricate (thats why you see a 64x64 in the assets folder as opposed to 16x16 which much better suits the current size)

- clean up a really ugly method in Hero (setSheets for sprite sheet animation breakdown) that was not generalizible at all and relied on hard-coded idiosyncacries. we wanted to make a "SheetHelper" class that took a Texture and an idiosyncacries[][] array and framesize/framewidth parameters (and other general things that made sheet break down much more generalized) and dealt with the texture with those idiosyncacries. again, we figured there was not enough time to expand the game to that point anyways and just made sure that our enemy sprite didnt have animations so we wouldnt have to rewrite that method again.

--------
to SEE CODE and RUN GAME : all the code is in master and the runnable game is GRAVITY_JUMP.jar.

to BUILD GAME : go to the build branch and run "sudo ./gradlew desktop:run", assuming youre on -nix. dont know what the equiv command is for windows but it probaaably has something to do with gradlew.bat. 

to build the jar yourself : "sudo ./gradlew desktop:dist" then the jar is in desktop/build/libs/desktop-1.0.jar



changes.txt
diff --git a/core/src/com/mygdx/gravity/Enemy.java b/core/src/com/mygdx/gravity/Enemy.java
index 876490d..1b2a8c0 100644
--- a/core/src/com/mygdx/gravity/Enemy.java
+++ b/core/src/com/mygdx/gravity/Enemy.java
@@ -1,5 +1,5 @@
 package com.mygdx.gravity;
-
+import java.util.*;
 import com.badlogic.gdx.math.Rectangle;
 import com.badlogic.gdx.graphics.g2d.TextureRegion;
 import com.badlogic.gdx.graphics.Texture;
@@ -11,17 +11,33 @@ public class Enemy {
 	TextureRegion defImage;
 	int damage;
 	int give;
+	float velocity;
+	float acceleration = 50f;
+	float slowdown = 0.000002f; 
+	//how long have we been moving in either left/right or up/down. 
+	//every time move in the same direction, add 1 (up/right), or -1 (left/down). 
+	//every time change, set to 1 (up/right) or -1 (left/down)
+
 	boolean alive; 
 
 	public final static String DEFAULT_IMAGE = "enemy.png";
 	public final static float DEFAULT_SIZE = 16;
-	public final static float SCALE = 4.5f;
-	public final static int MOVE_AMT = 100;
+	public final static int MIN_DMG = 5;
+	public final static int MAX_DMG = 10;
+	public final static int MIN_GIVE = 50;
+	public final static int MAX_GIVE = 100;
+	public final static float SCALE = 3.5f;
+	public final static float MOVE_AMT_X = 3f;
+	public final static float MOVE_AMT_Y = 2f;
+	public final static int MAX_X = 0;
+	public final static int MAX_Y = 0;
+
 	public Enemy(float spawnX, float spawnY, int dmg, int give) {
 		//It takes a nation of millions to hold us back.
 		this.rectRep = new Rectangle(spawnX, spawnY, SCALE * DEFAULT_SIZE, SCALE * DEFAULT_SIZE);
 		this.defImage = new TextureRegion(new Texture(Gdx.files.internal(DEFAULT_IMAGE)));
-		this.damage = damage;
+		this.damage = dmg;
+		
 		this.give = give;
 		this.alive = true;
 	}
@@ -32,24 +48,27 @@ public class Enemy {
 		this.defImage = null;
 	}
 
-	public void floatLeft(int amt) {
-		this.rectRep.x -= amt;
-		floatY(amt);	
-	}
+	public void move(float amtX, float amtY, Hero hero) {
+		//will angrily shake while trying to reach hero
+		this.rectRep.x += amtX;
+		this.rectRep.y += amtY;
+		this.velocity += this.acceleration;
+		this.rectRep.x += slowdown * Math.random() * velocity * (hero.rectRep.x - this.rectRep.x);
+		this.rectRep.y += slowdown * Math.random() * velocity * (hero.rectRep.y - this.rectRep.y);
 
-	public void floatRight(int amt) {
-		this.rectRep.x += amt;
-		floatY(amt);
 	}
 
-	public void floatY(int amt) {
-		int direction = randrange(-1,2);
-		this.rectRep.y += direction * Math.log(amt) / Math.log(2);
-	}
+	public boolean collide(Platform platform) {
+		if (this.collide(platform.rectRep)) {
 
-	public int randrange(int low, int high) {
-		//[low, high)
-		return low + (int) Math.floor(Math.random() * (high - low));
+			return true;
+		} else{
+			return false;
+		}
 	}
-	
+	public boolean collide(Rectangle object) {
+		boolean intersects = (this.rectRep.x <= object.x + object.width && this.rectRep.x + this.rectRep.width >= object.x && 
+					  this.rectRep.y <= object.y + object.height && this.rectRep.y + this.rectRep.height >= object.y);
+		return intersects;
+	}	
 }
\ No newline at end of file
diff --git a/core/src/com/mygdx/gravity/Hero.java b/core/src/com/mygdx/gravity/Hero.java
index e1ba3d4..60d0aca 100644
--- a/core/src/com/mygdx/gravity/Hero.java
+++ b/core/src/com/mygdx/gravity/Hero.java
@@ -13,11 +13,12 @@ public class Hero {
 	boolean isJumping;
 	int vDirection;
 	int hDirection;
-	int coin; 
+	int coin = HEALTH; 
 	int curr = 0;
 	long sinceLast;
 	float velocity = INIT_V;
 
+	public static final int HEALTH = 1000;
 	public static final int FEET_UP = 1;
 	public static final int FEET_DOWN = -1;
 	public static final String DEFAULT_MOVE_LEFT_IMAGE = "leftmove.png";
@@ -28,11 +29,11 @@ public class Hero {
 	public static final int DEFAULT_IMG_HEIGHT = 31;
 	public static final float SCALE = 4.5f;
 	public static final float FRAMES_PER_SECOND = 0.12f;
-	public static final float GRAVITY = 25f;
+	public static final float GRAVITY = 30f;
 	public static final float INIT_V = 0;
 	public static final int LEFT = 1;
 	public static final int RIGHT = 0;
- 	public static final int MOVE_AMT = 300;
+ 	public static final int MOVE_AMT = 600;
 
 	public Hero(float x, float y, float imgX, float imgY, float width, float height, 
 				Texture moveLeftSheet, Texture moveRightSheet) {
@@ -50,33 +51,6 @@ public class Hero {
 			new Texture(Gdx.files.internal(DEFAULT_MOVE_RIGHT_IMAGE)));
 	}
 
-	public TextureRegion[][] setSheets(Texture moveLeftSheet, Texture moveRightSheet) {
-		//breaks sprite sheets into Animations.
-		TextureRegion[] moveLeft = new TextureRegion[12];
-		TextureRegion[] moveRight = new TextureRegion[12];
-		TextureRegion[] moveLeftFlipped = new TextureRegion[12];
-		TextureRegion[] moveRightFlipped = new TextureRegion[12];
-		int l_startX = 0;
-		int r_endX = 0; //idiosyncracies in these specific frames
-		for (int i = 0; i < 12; i++) {
-			if (i >= 2 && i <= 4) l_startX = -1; //idiosyncracies in this specific frame 
-			if (i == 3) r_endX = 1; 
-			moveLeft[i] = new TextureRegion(moveLeftSheet, 32 * i + DEFAULT_START_X + l_startX, DEFAULT_START_Y,
-				DEFAULT_IMG_WIDTH + -1 * l_startX, DEFAULT_IMG_HEIGHT);
-
-			moveLeftFlipped[i] = new TextureRegion(moveLeft[i]);
-			moveLeftFlipped[i].flip(false, true);
-			moveRight[i] = new TextureRegion(moveRightSheet, 32 * i + DEFAULT_START_X + r_endX, DEFAULT_START_Y,
-				DEFAULT_IMG_WIDTH + r_endX, DEFAULT_IMG_HEIGHT);
-
-			moveRightFlipped[i] = new TextureRegion(moveRight[i]);
-			moveRightFlipped[i].flip(false, true);
-		}
-
-		TextureRegion[][] answer = {moveLeft, moveRight, moveLeftFlipped, moveRightFlipped}; 
-		return answer;
-	}
-
 	public void chooseTexture() {
 			this.currImage = this.textureRegions[this.vDirection + this.hDirection + 1][curr];
 	}
@@ -108,6 +82,10 @@ public class Hero {
  			sinceLast = 0;
 		}
 	}
+
+	public void moveVert(float amt) {
+		this.rectRep.y += amt;
+	}
 	public void stabilize() {
 		this.curr = 0;

core/src/com/mygdx/gravity/Hero.java
package com.mygdx.gravity;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.utils.*;

public class Hero {
	Rectangle rectRep;
	TextureRegion[][] textureRegions; //moveLeft, moveRight, moveLeftFlipped, moveRightFlipped 
	TextureRegion currImage; //what to render right now?
	boolean isJumping;
	int vDirection; //direction youre moving up/down
	int hDirection; //direction youre move left/right
	int coin = HEALTH; //HEALTH being your original wallet amt. 
	int curr = 0; //current frame index in the textureRegion[] index. 
	long sinceLast;
	float velocity = INIT_V;

	public static final int HEALTH = 1000;
	public static final int FEET_UP = 1; //vertical direction
	public static final int FEET_DOWN = -1;
	public static final String DEFAULT_MOVE_LEFT_IMAGE = "leftmove.png";
	public static final String DEFAULT_MOVE_RIGHT_IMAGE = "rightmove.png";
	public static final int DEFAULT_START_X = 7; //7 pixels of nothing on the 32x32
	public static final int DEFAULT_START_Y = 1; //from top-to-bottom, we want to cut off the extra pixel on top of his head
	public static final int DEFAULT_IMG_WIDTH = 18; //extra 7 pixels of nothing on the right and left : 32 - 7 * 2
	public static final int DEFAULT_IMG_HEIGHT = 31;
	public static final float SCALE = 4.5f;
	public static final float FRAMES_PER_SECOND = 0.12f; //how much time to spend on the whole set of frames
	public static final float GRAVITY = 30f;
	public static final float INIT_V = 0;
	public static final int LEFT = 1;
	public static final int RIGHT = 0;
 	public static final int MOVE_AMT = 600;

	public Hero(float x, float y, float imgX, float imgY, float width, float height, 
				Texture moveLeftSheet, Texture moveRightSheet) {
		this.rectRep = new Rectangle(x, y, SCALE * width, SCALE * height);
		this.textureRegions = setSheets(moveRightSheet, moveLeftSheet);
		this.currImage = textureRegions[LEFT][0]; //choosing moveLeft specifically is arbitrary
		this.isJumping = false;
		this.vDirection = -1;
		this.sinceLast = System.nanoTime();
	}

	public Hero(float spawnX, float spawnY) {
		this(spawnX, spawnY, DEFAULT_START_X, DEFAULT_START_Y, DEFAULT_IMG_WIDTH, DEFAULT_IMG_HEIGHT, 
			new Texture(Gdx.files.internal(DEFAULT_MOVE_LEFT_IMAGE)), 
			new Texture(Gdx.files.internal(DEFAULT_MOVE_RIGHT_IMAGE)));
	}

	public void chooseTexture() {
			this.currImage = this.textureRegions[this.vDirection + this.hDirection + 1][curr];
	}
	public void animate() {
		float slInSecs = (System.nanoTime() - sinceLast) / (float) (Math.pow(10,9)); //time since last frame in secs
 		float timePerFrame = Hero.FRAMES_PER_SECOND / textureRegions[LEFT].length; 
 		//specifically choosing moveleft is arbitrary, since theyre all the same : 12 frames 
 		if (slInSecs > timePerFrame) {
 			curr = ((int)(curr + slInSecs / timePerFrame)) % textureRegions[LEFT].length;
 			//find the difference in time relative to how much time you should spend in time per frame,
 			//add curr to that, then the mod takes care of looping from 0-12. 
 			sinceLast = System.nanoTime(); //update sinceLast
 		}
	}
	public void moveLeft(float amt, boolean animate) {
		this.hDirection = LEFT;
		this.rectRep.x -= amt;
		if (animate) animate(); //whether you should animate the movement or if the character is mid-jump 
		else stabilize();
	}

	public void moveRight(float amt, boolean animate) {
		this.hDirection = RIGHT;
		this.rectRep.x += amt;
		if (animate) animate();
		 else stabilize();
	}

	public void moveVert(float amt) {
		this.rectRep.y += amt;
	}
	public void stabilize() {
		//stop moving!
		this.curr = 0; //reset animation indices
		this.sinceLast = 0;
	}

	//character's "jump" is really flipping up and down until it hits a platform of some sort, much like Gravity Guy
	public void jump() {
		this.isJumping = true;
		this.vDirection = -1 * vDirection;
		this.getRect().y += 20 * vDirection;
	}

	public float jumpIter(float accel) {
		if (this.isJumping) this.velocity += this.vDirection * accel;
		return this.velocity;		
	}

	public void gainCoin(int amt) {
		this.coin += amt;
	}

	public void loseCoin(int amt) {
		this.coin -= amt;
	}

	public boolean collide(Rectangle object) {
		boolean intersects = (this.rectRep.x <= object.x + object.width && this.rectRep.x + this.rectRep.width >= object.x && 
					  this.rectRep.y <= object.y + object.height && this.rectRep.y + this.rectRep.height >= object.y);
		return intersects;
	}


	public boolean collide(Enemy enemy) {
		boolean loseCoin  = true;
		if (this.collide(enemy.rectRep)) {
			if ((vDirection == FEET_UP && rectRep.y < enemy.rectRep.y) || 
				(vDirection == FEET_DOWN && enemy.rectRep.y < rectRep.y))
				loseCoin = false;
			if (loseCoin) loseCoin(enemy.damage);
			else { 
				gainCoin(enemy.give);
				this.jump();
				enemy.die();
			}

		}
		return loseCoin;
	}
 	public boolean collide(Platform platform) {
		if (this.collide(platform.rectRep)) {
			this.isJumping = false;
			if (this.vDirection == FEET_UP) {
				this.rectRep.y = platform.rectRep.y - this.rectRep.height;
			} else {
				this.rectRep.y = platform.rectRep.y + platform.rectRep.height;
			}
			this.velocity = INIT_V;
			return true;
		} else{
			return false;
		}
	}

	public Rectangle getRect() {
		//GET WRECKED! 
		return this.rectRep;
	}

	public TextureRegion[][] setSheets(Texture moveLeftSheet, Texture moveRightSheet) {
		//breaks sprite sheet textures into 2darray with flipped sheets as well

		//this whole method breaks my heart. thats why i put it at the bottom. ugly piece of shit. no one likes you. youre at the bottom,
		//like an untouchable in the ancient indian caste system.
		// i wish i had more time to implement something more generalizible.
		//but this game simply doesnt have that scope and im late as it is. 
		TextureRegion[] moveLeft = new TextureRegion[12];
		TextureRegion[] moveRight = new TextureRegion[12];
		TextureRegion[] moveLeftFlipped = new TextureRegion[12];
		TextureRegion[] moveRightFlipped = new TextureRegion[12];
		int l_startX = 0;
		int r_endX = 0; //idiosyncracies in these specific frames
		for (int i = 0; i < 12; i++) {
			if (i >= 2 && i <= 4) l_startX = -1; //idiosyncracies in this specific frame 
			if (i == 3) r_endX = 1; 
			moveLeft[i] = new TextureRegion(moveLeftSheet, 32 * i + DEFAULT_START_X + l_startX, DEFAULT_START_Y,
				DEFAULT_IMG_WIDTH + -1 * l_startX, DEFAULT_IMG_HEIGHT);

			moveLeftFlipped[i] = new TextureRegion(moveLeft[i]);
			moveLeftFlipped[i].flip(false, true);
			moveRight[i] = new TextureRegion(moveRightSheet, 32 * i + DEFAULT_START_X + r_endX, DEFAULT_START_Y,
				DEFAULT_IMG_WIDTH + r_endX, DEFAULT_IMG_HEIGHT);

			moveRightFlipped[i] = new TextureRegion(moveRight[i]);
			moveRightFlipped[i].flip(false, true);
		}

		TextureRegion[][] answer = {moveLeft, moveRight, moveLeftFlipped, moveRightFlipped}; 
		return answer;
	}
}
